tpu/templates/tpu_cnn_estimator/trainer.py

```diff
@@ -0,0 +1,180 @@
+    max_pool_1 = tf.layers.max_pooling2d(conv_1, pool_size=(2, 2), strides=(2, 2))
+
+    batch_size = features.shape[0]
+    flattened = tf.reshape(max_pool_1, (batch_size, -1))
+    logits = tf.layers.dense(flattened, NUM_CLASSES)
+
+    predictions = tf.multinomial(logits, num_samples=1)
+    loss = None
+    train_op = None
+
+    if mode == tf.estimator.ModeKeys.TRAIN:
+        # define loss
+        loss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits))
+
+        # define train_op
+        optimizer = tf.train.RMSPropOptimizer(learning_rate=0.05)
+
+        # wrapper to make the optimizer work with TPUs
+        if params['use_tpu']:
+            optimizer = tf.contrib.tpu.CrossShardOptimizer(optimizer)
+
+        train_op = optimizer.minimize(loss, global_step=global_step)
+
+    if params['use_tpu']:
+        # TPU version of EstimatorSpec
+        return tf.contrib.tpu.TPUEstimatorSpec(
+            mode=mode,
+            predictions=predictions,
+            loss=loss,
+            train_op=train_op
+        )
+    else:
+        return tf.estimator.EstimatorSpec(
+            mode=mode,
+            predictions=predictions,
+            loss=loss,
+            train_op=train_op
+        )
+
+
+def train_input_fn(params={}):
+    # make some fake image classification data
+    data_size = 100
+    x = np.random.rand(data_size, 28, 28, 1)
+    y = np.random.randint(0, NUM_CLASSES, data_size)
+
+    x_tensor = tf.constant(x, dtype=tf.float32)
+    y_tensor = tf.constant(y, dtype=tf.int32)
+    dataset = tf.data.Dataset.from_tensor_slices((x_tensor, y_tensor))
+    dataset = dataset.repeat()
+
+    # TPUEstimator passes params when calling input_fn
+    batch_size = params.get('train_batch_size', 16)
+    dataset = dataset.batch(batch_size, drop_remainder=True)
+
+    # TPUs need to know all dimensions when the graph is built
+    # Datasets know the batch size only when the graph is run
+    def set_shapes(features, labels):
+        features_shape = features.get_shape().merge_with([batch_size, None, None, None])
+        labels_shape = labels.get_shape().merge_with([batch_size,])
+
+        features.set_shape(features_shape)
+        labels.set_shape(labels_shape)
+
+        return features, labels
+
+    dataset = dataset.map(set_shapes)
+    dataset = dataset.prefetch(tf.contrib.data.AUTOTUNE)
+
+    return dataset
+
+
+def main(args):
+    # pass the args as params so the model_fn can use
+    # the TPU specific args
+    params = vars(args)
+
+    if args.use_tpu:
+        # additional configs required for using TPUs
+        tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(args.tpu)
+        tpu_config = tf.contrib.tpu.TPUConfig(
+            num_shards=8, # using Cloud TPU v2-8
+            iterations_per_loop=args.save_checkpoints_steps
+        )
+
+        # use the TPU version of RunConfig
+        config = tf.contrib.tpu.RunConfig(
```

Comment:
It's interesting that even with tpu_config there is a separate RunConfig for TPU. This is more of a TPU/TF question but curious why is that?

Reply:
Actually answering my own question: I see it's due to the tpu_cluster_resolver